### PR TITLE
AJ-1284: Display a progress message when WDS app is in `UPDATING` state.

### PR DESCRIPTION
**Pull request by jladieu**
This is a re-apply of: https://github.com/DataBiosphere/terra-ui/pull/4211, updated to include a fix for a regression that change caused.

This addresses a serialization error on the WDS URL while reading the `V2Apps` response from Leo, which caused errors while attempting to load schema information using the an undefined proxy URL.

### Testing strategy
- [x] Updated existing unit tests to replicate the regression that was noticed in dev/staging/alpha.
- [x] Re-tested using a bee to both repro the original problem and confirm the fix.

### CODE REVIEW

**Comment by jladieu** [comment]
Note to reviewers: the first commit has already been reviewed, so unless you want to rehash it you can skim. However, if you want to see the code that caused the bug, look there!  I added some comments to aid in a game of spot-the-bug.

The fix is in the second commit.

File: src/data/Data.js [hunk 1]
```diff
@@ -577,9 +577,10 @@ export const WorkspaceData = _.flow(
     const region = isAzureWorkspace ? storageDetails.azureContainerRegion : storageDetails.googleBucketLocation;
 
     const wdsDataTableProvider = useMemo(() => {
-      const proxyUrl = !!wdsProxyUrl && wdsProxyUrl.state;
+      const app = wdsApp?.state;
+      const proxyUrl = app?.proxyUrl;
```
Comment (jladieu) [inline src/data/Data.js]: oopsie 1: `proxyUrl` isn't actually a field on app here, so this would always be undefined

File: src/data/Data.js [hunk 2]
```diff
@@ -680,24 +692,27 @@ export const WorkspaceData = _.flow(
...
-      if (!wdsProxyUrl || !['Ready', 'Error'].includes(wdsProxyUrl.status)) {
-        const wdsUrl = await loadWdsUrl(workspaceId);
-        if (wdsUrl) {
-          await loadWdsTypes(wdsUrl, workspaceId);
+      if (!wdsApp || !['Ready', 'Error'].includes(wdsApp.status)) {
+        const wdsApp = await loadWdsApp(workspaceId);
+        // TODO: figure out how not to make this redundant fetch, per
+        // https://github.com/DataBiosphere/terra-ui/pull/4202#discussion_r1319145491
+        if (wdsApp) {
+          loadWdsTypes(wdsApp.proxyUrl);
```
Comment (jladieu) [inline src/data/Data.js]: oopsie 1 again...proxyUrl isn't actually a field on app here, so this would always be undefined

Comment (jladieu) [inline src/data/Data.js]: oopsie 2: I stopped sending `workspaceId` 🤦

**Review by nawatts — APPROVED** [review]
I see what you meant here about how TS would have made this easier.

File: src/data/Data.js [hunk 3]
```diff
@@ -577,9 +577,10 @@ export const WorkspaceData = _.flow(
     const region = isAzureWorkspace ? storageDetails.azureContainerRegion : storageDetails.googleBucketLocation;
 
     const wdsDataTableProvider = useMemo(() => {
-      const proxyUrl = !!wdsProxyUrl && wdsProxyUrl.state;
+      const app = wdsApp?.state;
```
Comment (calypsomatic) [inline src/data/Data.js:580]: should this be called `appState`?

**Review by calypsomatic — APPROVED** [review]

File: src/data/Data.js (hunk 3 above)
Comment (jladieu) [reply src/data/Data.js:580]: Confusingly, no...the `wdsApp` is a `useState` that has a 'status' indicating whether it's loaded or not, and a `state` which is the stateful contents and happens to be an app.

**Review by davidangb — APPROVED** [review]
:+1: I ran this locally, and did not see the error I saw in the last version. Thank you!

File: src/data/Data.test.ts [hunk 4]
```diff
@@ -341,7 +371,8 @@ describe('WorkspaceData', () => {
 
     // Assert
     expect(mockListAppsV2).toHaveBeenCalledTimes(3); // third call, now running
-    expect(mockGetSchema).toHaveBeenCalled(); // fetch schema once running
+    expect(mockListAppsV2).toHaveBeenCalledWith('test-workspace-id'); // it should have been called with the correct ID
```
Comment (jladieu) [inline src/data/Data.test.ts:374]: Adding a comment to cross-ref from another PR to demonstrate the importance of verifying for specific args vs. verifying 'any' or loose args.

In this case, my 'loose' verify allowed a bug to creep through where the wrong arg signature was being passed through.

**Pull request merged**